### ISSUE TITLE
[P1][S00][statistics] Account for second-stage class cap in case-control inclusion weights

### ISSUE THREAD

**Issue by SzeChunYiu** [SzeChunYiu/ccb-testbeam#1019]
## Finding

`S00` currently performs **two sequential sampling stages** for the implementation-consistency classifier, but `sampling_weight` only accounts for the first stage.

1. In `scripts/01_build_pulse_table_from_root.py`, case/control rows are retained with class-dependent probabilities (`keep_selected`, `keep_rejected`) and assigned inverse-probability weights `1/p_keep`.
2. After all runs are concatenated, `scan_raw()` applies a second classwise random cap:

```python
max_sample = int(config["ml_check"]["max_train_per_class"]) + int(
    config["ml_check"]["max_test_per_class"]
)
...
for selected_value, subset in ml_rows.groupby("selected"):
    n = min(len(subset), max_sample)
    capped.append(
        subset.sample(
            n=n,
            random_state=int(config["ml_check"]["random_seed"])
            + int(selected_value),
        )
    )
```

The retained rows keep the first-stage `sampling_weight`; there is no factor for the second-stage inclusion probability `n / N_before_cap`.

`run_ml_check()` subsequently describes/uses those weights as restoring population prevalence and passes them into prevalence-sensitive held-out metrics and fitting/calibration. When the cap binds, that interpretation is false.

## Deterministic counterexample from current repository counts/config

Current `S00` report/config imply:

- total events across the four configured groups = `409815 + 388879 + 35943 + 262091 = 1,096,728`;
- four analysed B-stave candidate rows/event => `4,386,912` population rows;
- selected pulses = `640,737`;
- rejected candidate rows = `3,746,175`;
- source-population selected prevalence = `640737 / 4386912 = 0.1460565` (~14.61%).

Current first-stage keep probabilities are 0.20 selected and 0.05 rejected, while `max_train_per_class=60000` and `max_test_per_class=30000`, so the second-stage cap is 90,000/class.

Expected first-stage retained counts are approximately:

- selected: `0.20 * 640737 = 128,147.4` (>90,000);
- rejected: `0.05 * 3,746,175 = 187,308.75` (>90,000).

So **both caps bind strongly**. With equal 90k class caps but stale weights 5 and 20, the weighted selected prevalence becomes

`(90000*5) / (90000*5 + 90000*20) = 0.20`,

not ~0.1461.

This numerical example is an expectation-level illustration because first-stage sampling is random; the production fix must use the **actual pre-cap class counts in that realization**.

## Why this matters scientifically

The exact S00 selected-pulse count (`CL-001`) is unaffected; this defect is confined to the bounded ML/implementation-consistency diagnostic.

Prevalence-sensitive quantities can be biased/misinterpreted when the stale weights are used, including weighted accuracy, Brier score, average precision, calibrated probabilities/intercept, and threshold-dependent summaries. ROC AUC is much less sensitive here because the omitted second-stage factor is constant within class; do not claim that all ranking metrics are invalidated.

## Four role-separated review passes

These are sequential adversarial roles, not independent human reviewers.

### 1. Statistical-sampling lead

The relevant inclusion probability is the product of sequential inclusion probabilities. For class `c`, if stage 1 retains with `p1_c`, the realized stage-1 sample contains `N1_c`, and the cap retains `n2_c=min(N1_c, cap)`, then for a retained row

`pi_c = p1_c * (n2_c / N1_c)`

and the design weight should be

`w_c = 1 / pi_c = (1/p1_c) * (N1_c/n2_c)`.

If the cap is inactive, the multiplier is exactly 1.

### 2. Adversarial reviewer

The strongest falsifier is the current equal-class cap itself. It deliberately changes the retained class mixture after the first sampling stage. Keeping only `1/p1_c` cannot generally reconstruct the source class prevalence. The 20.0% versus 14.61% calculation above proves that the present claim of population restoration is not algebraically closed under the current design.

Alternative hypothesis considered: perhaps the cap is merely computational and never binds. Current counts/config reject that alternative: expected first-stage samples exceed 90k for both classes by large margins.

### 3. Independent validation/statistics reviewer

The fix should be tested as a survey-sampling design property, independent of ROOT files. Construct a synthetic finite population with known class prevalence; apply class-dependent stage-1 Bernoulli sampling and the class cap; compare source prevalence against estimates made with (a) stale stage-1-only weights and (b) composite weights. Repeat across RNG seeds and require the composite estimator to be design-unbiased within Monte Carlo tolerance while the intentionally adversarial stale-weight setup fails.

Also test the exact realized accounting: `N_before_cap`, `N_after_cap`, and the cap probability used to construct every retained row's weight must be recorded.

### 4. Claims/provenance reviewer

Any S00 report/manifest wording saying the current weights “restore population prevalence” must remain gated until the composite design is implemented and verified. Preserve the distinction between the exact deterministic pulse-count gate and this diagnostic ML branch; do not downgrade `CL-001` because of this issue.

## Literature basis

Primary methodological references:

- Horvitz, D. G. & Thompson, D. J. (1952), *A Generalization of Sampling Without Replacement from a Finite Universe*, JASA 47(260), 663–685. DOI: `10.1080/01621459.1952.10483446`.
- Prentice, R. L. & Pyke, R. (1979), *Logistic disease incidence models and case-control studies*, Biometrika 66(3), 403–411. DOI: `10.1093/biomet/66.3.403`.
- Fithian, W. & Hastie, T. (2014), *Local case-control sampling: Efficient subsampling in imbalanced data sets*, Annals of Statistics 42(5), 1693–1724. DOI: `10.1214/14-AOS1220`.

The narrow point used here is not that S00 must adopt a particular case-control estimator; it is that a population-weight interpretation must account for the actual sampling design that produced the retained rows.

## Implementation plan

1. Factor the cap into a tested helper, e.g. `apply_class_cap_with_sampling_weights(...)`.
2. Before capping each class, record `N_before_cap` and `N_after_cap`.
3. Define `p_cap = N_after_cap / N_before_cap` (1 when inactive).
4. Multiply every retained row's existing first-stage weight by `1/p_cap`.
5. Prefer explicit provenance columns such as `stage1_sampling_probability`, `cap_sampling_probability`, `sampling_probability`, and `sampling_weight`, or machine-readable equivalent in the manifest.
6. Add a weighted-prevalence closure diagnostic against the known pre-subsampling population totals accumulated by `scan_raw()`.
7. Rerun the S00 implementation-consistency diagnostic and regenerate report/manifest outputs.
8. If the sklearn calibration path cannot accept the corrected weights, fail closed or label that result unweighted; do not silently fall back while retaining population-restoration language. (The compatibility fallback itself should be audited separately.)

## Required negative controls

- **Cap inactive:** second-stage multiplier is exactly 1 and existing first-stage weights are unchanged.
- **One class capped:** only that class receives the second-stage multiplier.
- **Both classes capped, unequal stage-1 rates:** stale weights demonstrably fail source-prevalence closure; composite weights restore it in design expectation.
- **Class labels swapped with corresponding probabilities:** accounting remains internally consistent.
- **Malformed cap/accounting state:** zero/negative class size, impossible `N_after > N_before`, nonfinite probabilities, or zero inclusion probability must fail closed.

## Acceptance criteria

- [ ] Every retained ML row has an inclusion probability/weight consistent with **all** sampling stages that actually selected it.
- [ ] Tests cover inactive, one-class, and both-class cap regimes.
- [ ] Synthetic repeated-sampling closure demonstrates design-unbiased source prevalence within preregistered Monte Carlo tolerance.
- [ ] The S00 manifest records realized pre-cap/post-cap class counts and effective inclusion probabilities.
- [ ] Weighted prevalence reported by the rerun agrees with the source-population prevalence within expected finite-sampling uncertainty.
- [ ] Report text distinguishes prevalence-sensitive calibrated metrics from class-ranking metrics.
- [ ] `CL-001` and the deterministic selected-pulse count remain unchanged unless an independent data-integrity issue changes them.

## Dependencies / execution status

No real-data access is required to implement and unit-test the accounting fix. A production S00 rerun on immutable ROOT inputs is required only to regenerate the final diagnostic metrics/artifacts.

## Follow-up atoms deliberately not folded into this issue

Two neighboring concerns should be reviewed separately to keep this issue atomic:

- hyperparameter selection currently appears to use an unweighted CV ROC-AUC objective while final fitting/evaluation uses sampling weights;
- the sklearn compatibility `TypeError` fallback can silently calibrate without `sample_weight`.

Those are distinct estimator/compatibility questions and should not block implementing the missing second-stage inclusion factor.

**Comment by SzeChunYiu** [comment]
Live-main revalidation after filing: the defect is still present at protected `main` commit `da8a3e9398384e87e8b99bf50d345fde2107b98b` (the post-#1017 state). `scan_raw()` still creates first-stage inverse-probability weights, then class-caps `ml_rows` with `subset.sample(...)` and concatenates the capped rows without changing `sampling_weight`. This confirms #1019 is not stale relative to the latest S00 gate-state merge.

**Comment by SzeChunYiu** [comment]
Atomic-universe deduplication review: this is substantively the same sampling-design defect already tracked in #958 (second class-specific cap creates an additional inclusion probability not reflected in the stored weight). I am preserving the stronger algebraic example here but closing this issue as a duplicate so implementation has one canonical leaf. Please move any useful acceptance details from this thread into #958 rather than implementing twice.
